CityPuzzle/Classes/AddPageLogic.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace CityPuzzle.Classes
+{
+    public class AddPageLogic
+    {
+        private int puzzleNr = -1;
+        private List<Lazy<Puzzle>> allPuzzles;
+        public List<Lazy<Puzzle>> SelectedPuzzles { get; private set; }
+
+        public event EventHandler OnNoPuzzlesFound;
+        public event EventHandler OnEndOfPuzzles;
+        public event EventHandler<OnPuzzleChangeEventArgs> OnPuzzleChange;
+        public class OnPuzzleChangeEventArgs : EventArgs { public string Name; public string ImgAdress; public string About; }
+
+        public AddPageLogic()
+        {
+            allPuzzles = Sql.ReadPuzzles();
+        }
+
+        public void ChangePuzzle(bool direction)
+        {
+            if (allPuzzles.Count == 0)
+            {
+                OnNoPuzzlesFound?.Invoke(this, EventArgs.Empty);
+            }
+            else if (SelectedPuzzles.Contains(allPuzzles[puzzleNr++]) && direction)
+            {
+                ChangePuzzle(true);
+            }
+            else if (SelectedPuzzles.Contains(allPuzzles[puzzleNr--]) && !direction)
```

Comment:
allPuzzles -- and even puzzleNr -1 can lead to indexOutOfBoundsException. Should add extra check before using such indexing